### ISSUE TITLE
High CPU in TrPoolWorker

### ISSUE THREAD

**Issue by alanmilinovic** [ntop/ntopng#3059]
I am confused, so many issues and discussions about very high CPU and all of them are resolved. On my RPi 4 with buster and latest version of ntopng, CPU is always very high. It is so high that it slows down entire RPi so that I cannot use other applications, it is freezing because of that.

Ntopng is very cool application and I would like to use it but unfortunately I cannot. Are there any improvements in that direction? What I did is I limited interface only to tun0, it helped but very little, system is still very unstable with CPU eaten by Ntopng.

**Comment by emanuele-f** [comment]
Hi, thank you for your interest in improving ntopng. I'm upgrading the raspberry package, please wait one hour and retry with the new version. After upgrading, can you post the output of `top -H -p $(pgrep ntopng)` to know which are the ntopng threads which are causing the high load? Does the high load occur only if the web interface is opened?

**Comment by alanmilinovic** [comment]
Hi, thank you for a fast response. Sure I will be more than happy to test new version and past output here. High load is there as well in the background, when web interface is not opened.

**Comment by emanuele-f** [comment]
There is currently an issue with the build, I will update this thread once the new package is ready.

**Comment by alanmilinovic** [comment]
No worries, just to be on the same page, this is what I used as a source to get package on my RPi.

```
deb http://packages.ntop.org/apt/buster_pi armhf/
deb http://packages.ntop.org/apt/buster_pi all/
```

**Comment by emanuele-f** [comment]
New package now available. You can upgrade with:

```
sudo apt-get update
sudo apt-get install ntopng
```

**Comment by alanmilinovic** [comment]
```
top - 19:49:02 up 13:12,  3 users,  load average: 1.86, 1.93, 2.51           Threads:  22 total,   0 running,  22 sleeping,   0 stopped,   0 zombie
%Cpu(s): 22.6 us,  9.0 sy,  0.1 ni, 47.4 id, 20.7 wa,  0.0 hi,  0.3 si,  0.0 MiB Mem :   3664.2 total,    569.9 free,   1319.3 used,   1774.9 buff/cache
MiB Swap:    100.0 total,     97.5 free,      2.5 used.   2099.0 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
30408 ntopng    20   0  409728 116828  28596 D  38.3   3.1   0:06.23 TrPool+ 30406 ntopng    20   0  409728 116828  28596 S  15.0   3.1   0:10.19 TrPool+
30407 ntopng    20   0  409728 116828  28596 S  12.3   3.1   0:12.39 TrPool+ 30409 ntopng    20   0  409728 116828  28596 D   1.7   3.1   0:03.11 second+
30419 ntopng    20   0  409728 116828  28596 D   0.3   3.1   0:00.95 tun0    30282 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:01.52 ntopng 
30311 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.01 ntopng  30312 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ntopng 
30313 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ntopng  30314 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ntopng 
30315 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ntopng  30316 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ntopng 
30317 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.03 resolv+ 30410 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 ht_sta+
30411 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 stats_+ 30412 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 minute+
30413 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 5min.l+ 30414 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 hourly+
30415 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 daily.+ 30416 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 housek+
30417 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 discov+ 30418 ntopng    20   0  409728 116828  28596 S   0.0   3.1   0:00.00 timese+
```

**Comment by emanuele-f** [comment]
Can you post a screenshot of the ntopng footer (the badges which indicate the number of hosts and flows)? Both CPU and memory usage seem fine. Are you sure that this is a cpu related issue?

**Comment by alanmilinovic** [comment]
I am not sure if it is CPU, but that is what I saw as a higher number for ntopng process.

![image](https://user-images.githubusercontent.com/1752983/69337663-b10cf980-0c61-11ea-992a-b627b214a8b9.png)

**Comment by emanuele-f** [comment]
So you are monitoring a small network, ntopng should use low resources. Usually freezing issues on linux occur when the memory is exhausted and system begins to swap

**Comment by alanmilinovic** [comment]
But memory and swap memory looks fine.

**Comment by emanuele-f** [comment]
The CPU load from the `top` output above does not explain the freeze issue. Please report the output of `top` and `cat /proc/meminfo` while the system is freezing.

**Comment by M9k** [comment]
Good morning,
I have the same problem on a VM (x86_64).
![Immagine](https://user-images.githubusercontent.com/9288402/70702830-39b20f00-1ccf-11ea-800f-da4169a9a5a1.png)

grep -v "^#" /etc/ntopng/ntopng.conf
--online-license-check
-i=tcp://192.168.X.X:5556
-i=tcp://192.168.X.X:5556
-m=10.0.0.0/8,192.168.0.0/16,172.16.0.0/12

Thanks

**Comment by emanuele-f** [comment]
@M9k do you have any news regarding the `top -H` output after the new changes as asked privately?

**Comment by M9k** [comment]
@emanuele-f :
   PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
151263 ntopng    20   0 1682848 764568  11416 S 28,0 43,0  38:59.10 ht_state_update
151264 ntopng    20   0 1682848 764568  11416 S 25,0 43,0  39:04.35 ht_state_update
151265 ntopng    20   0 1682848 764568  11416 S  2,3 43,0   3:16.55 housekeeping.lu
151269 ntopng    20   0 1682848 764568  11416 S  1,3 43,0   3:20.67 second.lua
  6371 root      20   0  261228   1352    724 R  1,0  0,1  56:05.85 python
  6398 root      20   0  482424   1376    776 S  1,0  0,1  36:44.65 python
151268 ntopng    20   0 1682848 764568  11416 S  1,0 43,0  22:45.72 discover.lua
151266 ntopng    20   0 1682848 764568  11416 S  0,3 43,0   3:03.20 stats_update.lu
173761 adm.mca+  20   0  162020   2300   1592 R  0,3  0,1   0:00.05 top
[...]

Version:
v.3.9.191212    [Enterprise/Professional build]
Platform:       x86_64
Edition:        Enterprise

**Comment by alanmilinovic** [comment]
I can see sometimes ntopng process to use 90% of CPU power for few seconds. Usually it is around 20% what is as well very high if you ask me. All that in idle mode.

**Comment by emanuele-f** [comment]
Thank you for the info. Can you try to disable all the flow scripts and see if the load goes back to acceptable levels?

![2019-12-20_10-17](https://user-images.githubusercontent.com/5488003/71244062-95048280-2309-11ea-8ef0-aeb52143c34d.png)

**Comment by alanmilinovic** [comment]
By clicking on Interface I get this:
 Script "/usr/share/ntopng/scripts/lua/if_stats.lua" returned an error:
/var/lib/ntopng/plugins/alert_endpoints/nagios.lua:1: unexpected symbol near '<\135>'

**Comment by emanuele-f** [comment]
Please upgrade to the latest ntopng package, then post here the output of `ntopng --version` if the problem still occurs.

**Comment by alanmilinovic** [comment]
pi@1raspberrypi:~ $ ntopng --version
v.3.9.191220    [Enterprise/Professional/Embedded build]
GIT rev:        dev:818a21d9eeb48d844b9e2169b9b0dcf0e9c454e0:20191220
Pro rev:        r2624
Built on:       Raspbian GNU/Linux 10 (buster)
System Id:      44066669499602D2
Platform:       armv7l
Edition:        Enterprise Embedded
License Type:   Time-Limited [Empty license file]
Validity:       Until Fri Dec 20 16:35:10 2019

**Comment by simonemainardi** [comment]
please show exactly where you click as I cannot reproduce

**Comment by alanmilinovic** [comment]
Interface button in top horizontal menu.

**Comment by simonemainardi** [comment]
Make sure ntopng has been correctly updated, the old process stopped and the new process executed.

**Comment by alanmilinovic** [comment]
I updated package with normal linux update statement. Not sure what you are reffering to. Will try to purge and install package again.

**Comment by alanmilinovic** [comment]
`sudo systemctl restart ntopng` helped. Will disable flow scripts and let you know about performance.

**Comment by alanmilinovic** [comment]
Still same usage, 23% of CPU.

**Comment by emanuele-f** [comment]
Even if the scripts are disabled there is an overhead since some operations are always performed. In 8b7d711a3674d2af92b59bf647c23c6f040de714 I've added an option to disable the invocation. Please wait one hour for the new package to be available, then upgrade and edit `/usr/share/ntopng/scripts/callbacks/interface/ht_state_update.lua`, setting `skip_user_scripts = true`. This will disable flow scripts invocation. Please check out the cpu usage after that and report the output of `top -H -p $(pgrep ntopng)`

**Comment by alanmilinovic** [comment]
Ok, waiting for new package.

**Comment by emanuele-f** [comment]
New improvements implemented in https://github.com/ntop/ntopng/commit/87c336a4aae9e2ea0c8410830b24b42fa94375dd . The new package is now available. Please try both with and without `skip_user_scripts = true`.

**Comment by alanmilinovic** [comment]
Now it is 2%, with or without skip user scripts. Thank you for a great work! Happy new year!

**Comment by orestis46** [comment]
Im also having high CPU (>95%) running ntopng version v.3.9.200313 in Raspberry Pi 3 running Stretch 9.11 with LInux kernel 4.19.66-v7+ for my LAN environment. 
 
The problem with high CPU occurs as soon as im opening a browser from anywhere in my LAN network to reach the GUI of ntopng and check the active flows section.

 top - 09:31:04 up 10 days, 10:45,  6 users,  load average: 5.56, 4.54, 2.85
Threads:  38 total,   6 running,  32 sleeping,   0 stopped,   0 zombie
%Cpu0  : 88.5 us,  9.8 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  1.6 si,  0.0 st
%Cpu1  : 90.8 us,  8.6 sy,  0.0 ni,  0.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  : 87.1 us, 11.2 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  1.7 si,  0.0 st
%Cpu3  : 94.8 us,  5.2 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :   948304 total,   118472 free,   534116 used,   295716 buff/cache
KiB Swap:  2097148 total,  1065980 free,  1031168 used.   252840 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND                                                                          
22078 ntopng    20   0  591232 204396  29164 R 63.5 21.6   6:24.77 ntopng                                                                           
22077 ntopng    20   0  591232 204396  29164 R 61.5 21.6   6:22.60 ntopng                                                                           
22076 ntopng    20   0  591232 204396  29164 R 50.3 21.6   6:24.49 ntopng                                                                           
22075 ntopng    20   0  591232 204396  29164 R 46.4 21.6   5:50.27 ntopng                                                                           
22074 ntopng    20   0  591232 204396  29164 R 44.7 21.6   6:25.11 ntopng                                                                           
22126 ntopng    20   0  591232 204396  29164 R 42.8 21.6  24:50.58 minute.lua                                                                       
22128 ntopng    20   0  591232 204396  29164 S  1.6 21.6   2:05.92 second.lua                                                                       
22140 ntopng    20   0  591232 204396  29164 S  0.7 21.6   0:29.07 inline ifid 4                                                                    
22129 ntopng    20   0  591232 204396  29164 S  0.3 21.6   0:00.66 stats_update.lu                                                                  
22137 ntopng    20   0  591232 204396  29164 S  0.3 21.6   0:39.21 inline ifid 1                                                                    
22139 ntopng    20   0  591232 204396  29164 S  0.3 21.6   0:23.73 inline ifid 3                                                                    
22028 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:05.30 ntopng                                                                           
22073 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:02.17 ntopng                                                                           
22079 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:00.13 resolveLoop                                                                      
22111 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:40.01 ht_state_update                                                                  
22112 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:39.98 ht_state_update                                                                  
22113 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:38.14 ht_state_update                                                                  
22114 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:39.88 ht_state_update                                                                  
22115 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:09.07 stats_update.lu                                                                  
22116 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:08.80 stats_update.lu                                                                  
22117 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:08.92 stats_update.lu                                                                  
22118 ntopng    20   0  591232 204396  29164 S  0.0 21.6   0:09.23 stats_update.lu                                                                  
22119 ntopng    20   0  591232 204396  29164 S  0.0 21.6   1:27.56 5min.lua                                                                         
22120 ntopng    20   0  591232 204396  29164 S  0.0 21.6   1:27.95 5min.lua                                                                         
22121 ntopng    20   0  591232 204396  29164 S  0.0 21.6   1:30.70 5min.lua                                                                         
22122 ntopng    20   0  591232 204396  29164 S  0.0 21.6   1:27.91 5min.lua
                                                                        
                                                            
If i leave the GUI with the section of active flows open, then the output of top remains above >95% and i can see that the CPU load (according to top)  rises up to 7-8 values which means there is a plenty of work for the 4 CPUs which has to be done).
      

After closing the active flows section in ntopng GUI the CPU usage drops:

top - 09:37:58 up 10 days, 10:52,  6 users,  load average: 0.48, 2.27, 2.50
Threads:  38 total,   0 running,  38 sleeping,   0 stopped,   0 zombie
%Cpu0  :  0.7 us,  0.7 sy,  0.0 ni, 98.3 id,  0.3 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu1  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  :  1.7 us,  0.3 sy,  0.0 ni, 97.7 id,  0.3 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu3  :  1.3 us,  3.0 sy,  0.0 ni, 95.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :   948304 total,   116948 free,   532636 used,   298720 buff/cache
KiB Swap:  2097148 total,  1065980 free,  1031168 used.   254320 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND                                                                          
22128 ntopng    20   0  588160 201208  29164 S  1.3 21.2   2:11.49 second.lua                                                                       
22137 ntopng    20   0  588160 201208  29164 S  0.7 21.2   0:41.12 inline ifid 1                                                                    
22133 ntopng    20   0  588160 201208  29164 S  0.3 21.2   0:33.03 timeseries.lua                                                                   
22138 ntopng    20   0  588160 201208  29164 S  0.3 21.2   0:25.97 inline ifid 2                                                                    
22139 ntopng    20   0  588160 201208  29164 S  0.3 21.2   0:24.76 inline ifid 3                                                                    
22140 ntopng    20   0  588160 201208  29164 S  0.3 21.2   0:30.23 inline ifid 4                                                                    
22028 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:05.31 ntopng                                                                           
22073 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:02.33 ntopng                                                                           
22074 ntopng    20   0  588160 201208  29164 S  0.0 21.2   7:51.57 ntopng                                                                           
22075 ntopng    20   0  588160 201208  29164 S  0.0 21.2   7:05.40 ntopng                                                                           
22076 ntopng    20   0  588160 201208  29164 S  0.0 21.2   7:50.83 ntopng                                                                           
22077 ntopng    20   0  588160 201208  29164 S  0.0 21.2   7:50.14 ntopng                                                                           
22078 ntopng    20   0  588160 201208  29164 S  0.0 21.2   7:52.04 ntopng                                                                           
22079 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:00.13 resolveLoop                                                                      
22111 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:41.33 ht_state_update                                                                  
22112 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:41.27 ht_state_update                                                                  
22113 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:39.36 ht_state_update                                                                  
22114 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:41.22 ht_state_update                                                                  
22115 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:09.41 stats_update.lu                                                                  
22116 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:09.17 stats_update.lu                                                                  
22117 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:09.27 stats_update.lu                                                                  
22118 ntopng    20   0  588160 201208  29164 S  0.0 21.2   0:09.58 stats_update.lu                                                                  
22119 ntopng    20   0  588160 201208  29164 S  0.0 21.2   1:29.67 5min.lua                                                                         
22120 ntopng    20   0  588160 201208  29164 S  0.0 21.2   1:30.23 5min.lua                                                                         
22121 ntopng    20   0  588160 201208  29164 S  0.0 21.2   1:32.91 5min.lua                                                                         
22122 ntopng    20   0  588160 201208  29164 S  0.0 21.2   1:30.58 5min.lua   

Is this a normal and expected behaviour ?

**Comment by emanuele-f** [comment]
The high cpu load you are reporting is linked to the ntopng gui not to the periodic thread as this original issue. Please open a new issue.
